### ISSUE TITLE
Sometimes a kill + relaunch is needed to see all notifications

### ISSUE THREAD

**Issue by BasThomas** [rnystrom/GitHawk#121]
Happened only once though; no idea what caused it. Refreshing multiple times didn't help.

Safari:
![img_2191](https://user-images.githubusercontent.com/4190298/28120035-274a2ade-6717-11e7-932d-9a58476f74e9.JPG)
Before:
![img_2192](https://user-images.githubusercontent.com/4190298/28120033-2613a6e0-6717-11e7-8808-fbe681517762.JPG)
After:
![img_2193](https://user-images.githubusercontent.com/4190298/28120032-2608b276-6717-11e7-8389-46a032b9bd21.JPG)

**Comment by rnystrom** [comment]
Ugh. Ya I'm not sure what this is from. It's either Apollo's cache or GitHub caching API responses. I nil out the HTTP cache, so that shouldn't be it, but I'm just not sure... Will keep an eye on it.

**Comment by rnystrom** [comment]
Ok here's to blame on issue/PR aggressive caching:

https://github.com/rnystrom/Freetime/commit/9ba1a51704e64fc35576d6a12ba684b5fc5166db

Doesn't explain notifications tho, that's a totally different client+API.

**Comment by BasThomas** [comment]
I am seeing this still / again... a lot. Refresh for a bunch of issues, going through all of them and marking them as read. Then when I go to github.com, I see some issues as unread (others than previously marked as read). Need to kill and relaunch the app to see them :/

**Comment by rnystrom** [comment]
@BasThomas since you're seeing this so often, kind digging into it a little?

**Comment by BasThomas** [comment]
I'll see what I can do! 👍

**Comment by BasThomas** [comment]
- "Mark all as read" button is still enabled w/out the issues showing up
- The issues in question are also not found in the "read" section

**Comment by BasThomas** [comment]
- mark as read does work

**Comment by BasThomas** [comment]
- does not resurface them as read until restart

**Comment by rnystrom** [comment]
Now that I have badges enabled, I'm noticing this now too. Very frustrating. Not sure if Alamofire is doing something special w/ the etag or what.

**Comment by rnystrom** [comment]
Copied the curl request via FLEX when I had a refresh bug (replaced my auth token):

```
curl -v -X GET 'https://api.github.com/notifications?access_token=123abc&all=true&page=1&participating=false&per_page=100' -H 'User-Agent: Freetime/1.10.0 (com.whoisryannystrom.freetime; build:1897; iOS 11.0.0) Alamofire/4.4.0' -H 'Accept-Language: en-US;q=1.0' -H 'Authorization: Bearer 123abc' -H 'Accept: */*' -H 'Accept-Encoding: gzip;q=1.0, compress;q=0.5' -H 'Cookie:' 
```

Killed the app and restarted. It sent this request:

```
curl -v -X GET 'https://api.github.com/notifications?access_token=123abc&all=true&page=1&participating=false&per_page=100' -H 'User-Agent: Freetime/1.10.0 (com.whoisryannystrom.freetime; build:1897; iOS 11.0.0) Alamofire/4.4.0' -H 'Accept-Language: en-US;q=1.0' -H 'Authorization: Bearer 123abc' -H 'Accept-Encoding: gzip;q=1.0, compress;q=0.5' -H 'Cookie:' 
```

The only difference is the `-H 'Accept: */*'`. Could that really be it?

**Comment by BasThomas** [comment]
I guess it must be? Very weird. Can we open an issue at GitHub somewhere?

**Comment by BasThomas** [comment]
Or it’s some weird caching issue on our side still?

**Comment by rnystrom** [comment]
> Or it’s some weird caching issue on our side still?

This could totally be the case, tho we're still [explicitly niling the cache](https://github.com/rnystrom/GitHawk/blob/master/Classes/Systems/Alamofire%2BGithubAPI.swift#L34).

Also hard to repro b/c as soon as I relaunch, the response is fresh. Maybe I can spend some time using my test account to blast notifications and wait until I start seeing stale data on refresh?

**Comment by BasThomas** [comment]
At this point, I guess just trying things is a good idea? It’s quite a nasty and impactful bug.

**Comment by BasThomas** [comment]
Thinking out loud: I think the “mark as read” is still available if this happens? Or not? If so, maybe we can do some logging / force reload when the amount of notifications in Unread is 0 but the mark all is still available?

**Comment by rnystrom** [comment]
@BasThomas the "Mark All" button `isEnabled` state is driven from the [same data that powers the data source](https://github.com/rnystrom/GitHawk/blob/master/Classes/Notifications/NotificationsViewController.swift#L81-L82), so I don't think that's it. I hate this bug, but its super visible and looks bad. I honestly need to just set aside some time and hammer on it.

**Comment by BasThomas** [comment]
Yeah - just confirmed that it’s not working with the mark as read.

**Comment by rnystrom** [comment]
Ok this is _really_ weird. The background job says 9 new. I foreground + refresh and it **adds a new notification** but limits it to 6. So I'm receiving **new** notifications, even tho they are technically stale. Attaching debugger and seeing what's up.

**Comment by BasThomas** [comment]
Yesssssssss

**Comment by BasThomas** [comment]
Oh noes - I have a notification on GH.com now, but it is not at all showing in the app - even after kill+relaunch :(. Not sure if this is different, as a relaunch doesn't fix? Want me to open another issue?
![img_1551](https://user-images.githubusercontent.com/4190298/30971565-898a652c-a468-11e7-8c0c-0f8420dd1737.PNG)
![img_1552](https://user-images.githubusercontent.com/4190298/30971566-89a5f71a-a468-11e7-8bff-1e20ad7c27a7.PNG)

**Comment by BasThomas** [comment]
Seems like it was a push merely for file changes - do we filter these (on purpose) maybe?

**Comment by rnystrom** [comment]
API doesn’t report them

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BasThomas** [comment]
What do you mean? Think it’s in the response?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rnystrom** [comment]
@BasThomas the [notifciations API](https://developer.github.com/v3/activity/notifications/#list-your-notifications) only reports for [these reasons](https://developer.github.com/v3/activity/notifications/#notification-reasons) on an issue, PR, or commit. The API wont ever deliver notifications for things like "new release pushed". Your "file changes" notification wont be covered either.

**Comment by BasThomas** [comment]
Sorry - I don't think I don't understand. This was in the response:

```
{
    "unread" : true,
    "reason" : "subscribed",
    "id" : "261743420",
    "repository" : {
      "private" : false,
      "downloads_url" : "https://api.github.com/repos/Moya/Moya/downloads",
      "issues_url" : "https://api.github.com/repos/Moya/Moya/issues{/number}",
      "archive_url" : "https://api.github.com/repos/Moya/Moya/{archive_format}{/ref}",
      "id" : 23013268,
      "tags_url" : "https://api.github.com/repos/Moya/Moya/tags",
      "trees_url" : "https://api.github.com/repos/Moya/Moya/git/trees{/sha}",
      "branches_url" : "https://api.github.com/repos/Moya/Moya/branches{/branch}",
      "notifications_url" : "https://api.github.com/repos/Moya/Moya/notifications{?since,all,participating}",
      "pulls_url" : "https://api.github.com/repos/Moya/Moya/pulls{/number}",
      "releases_url" : "https://api.github.com/repos/Moya/Moya/releases{/id}",
      "html_url" : "https://github.com/Moya/Moya",
      "forks_url" : "https://api.github.com/repos/Moya/Moya/forks",
      "url" : "https://api.github.com/repos/Moya/Moya",
      "description" : "Network abstraction layer written in Swift.",
      "hooks_url" : "https://api.github.com/repos/Moya/Moya/hooks",
      "stargazers_url" : "https://api.github.com/repos/Moya/Moya/stargazers",
      "issue_comment_url" : "https://api.github.com/repos/Moya/Moya/issues/comments{/number}",
      "assignees_url" : "https://api.github.com/repos/Moya/Moya/assignees{/user}",
      "languages_url" : "https://api.github.com/repos/Moya/Moya/languages",
      "full_name" : "Moya/Moya",
      "subscribers_url" : "https://api.github.com/repos/Moya/Moya/subscribers",
      "comments_url" : "https://api.github.com/repos/Moya/Moya/comments{/number}",
      "merges_url" : "https://api.github.com/repos/Moya/Moya/merges",
      "contributors_url" : "https://api.github.com/repos/Moya/Moya/contributors",
      "name" : "Moya",
      "issue_events_url" : "https://api.github.com/repos/Moya/Moya/issues/events{/number}",
      "commits_url" : "https://api.github.com/repos/Moya/Moya/commits{/sha}",
      "git_commits_url" : "https://api.github.com/repos/Moya/Moya/git/commits{/sha}",
      "keys_url" : "https://api.github.com/repos/Moya/Moya/keys{/key_id}",
      "labels_url" : "https://api.github.com/repos/Moya/Moya/labels{/name}",
      "git_refs_url" : "https://api.github.com/repos/Moya/Moya/git/refs{/sha}",
      "blobs_url" : "https://api.github.com/repos/Moya/Moya/git/blobs{/sha}",
      "owner" : {
        "id" : 13662162,
        "organizations_url" : "https://api.github.com/users/Moya/orgs",
        "received_events_url" : "https://api.github.com/users/Moya/received_events",
        "following_url" : "https://api.github.com/users/Moya/following{/other_user}",
        "login" : "Moya",
        "avatar_url" : "https://avatars3.githubusercontent.com/u/13662162?v=4",
        "url" : "https://api.github.com/users/Moya",
        "subscriptions_url" : "https://api.github.com/users/Moya/subscriptions",
        "type" : "Organization",
        "repos_url" : "https://api.github.com/users/Moya/repos",
        "html_url" : "https://github.com/Moya",
        "events_url" : "https://api.github.com/users/Moya/events{/privacy}",
        "site_admin" : false,
        "starred_url" : "https://api.github.com/users/Moya/starred{/owner}{/repo}",
        "gists_url" : "https://api.github.com/users/Moya/gists{/gist_id}",
        "gravatar_id" : "",
        "followers_url" : "https://api.github.com/users/Moya/followers"
      },
      "subscription_url" : "https://api.github.com/repos/Moya/Moya/subscription",
      "deployments_url" : "https://api.github.com/repos/Moya/Moya/deployments",
      "milestones_url" : "https://api.github.com/repos/Moya/Moya/milestones{/number}",
      "contents_url" : "https://api.github.com/repos/Moya/Moya/contents/{+path}",
      "events_url" : "https://api.github.com/repos/Moya/Moya/events",
      "teams_url" : "https://api.github.com/repos/Moya/Moya/teams",
      "collaborators_url" : "https://api.github.com/repos/Moya/Moya/collaborators{/collaborator}",
      "git_tags_url" : "https://api.github.com/repos/Moya/Moya/git/tags{/sha}",
      "statuses_url" : "https://api.github.com/repos/Moya/Moya/statuses/{sha}",
      "fork" : false,
      "compare_url" : "https://api.github.com/repos/Moya/Moya/compare/{base}...{head}"
    },
    "subscription_url" : "https://api.github.com/notifications/threads/261743420/subscription",
    "subject" : {
      "latest_comment_url" : null,
      "title" : "[WIP][CI] Add new script to exit on non-trivial changes (docs, markdown, etc)",
      "url" : "https://api.github.com/repos/Moya/Moya/pulls/1324",
      "type" : "PullRequest"
    },
    "last_read_at" : "2017-09-28T13:27:43Z",
    "updated_at" : "2017-09-28T13:28:28Z",
    "url" : "https://api.github.com/notifications/threads/261743420"
  },
```

**Comment by Sherlouk** [comment]
Reopening as I've had the same issue today. The network response had items but they weren't shown in the app. Forgot to save the log (sorry!)

**Comment by Sherlouk** [comment]
`latest_comment_url` is null, that's in a guard in `Subject.swift:16` - Prevents the notification response from being made.

Comment URL wouldn't exist if it's a code change?

**Comment by Sherlouk** [comment]
@rnystrom ^^ Opinion on this one?

**Comment by rnystrom** [comment]
@Sherlouk wouldn't that effect the background request _and_ the foreground request though?

Agree we can get rid of it, just unsure if that would cause inconsistencies since technically all requests would be broken until that notification goes away. Right?

**Comment by Sherlouk** [comment]
@rnystrom The issue I was having though is no matter what I did the notification wasn't showing. Regardless of background/foreground/how many times I kill and re-enter the app.

Which would insinuate that it is something just not being able to parse the notification

**Comment by rnystrom** [comment]
Ya that makes sense and we should def fix. So not part of the kill+relaunch bug where stuff is cached, more that stuff wasn't showing at all b/c of this.

Great find. PR would be awesome!

**Comment by BasThomas** [comment]
Reopening #939 then

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rnystrom** [comment]
Oops didn’t mean to close

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rnystrom** [comment]
Dang reference events are working 🤔

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BasThomas** [comment]
?

**Comment by rnystrom** [comment]
All I see on this thread is a closed event, no reference from the PR, thought I accidentally closed instead of auto.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BasThomas** [comment]
Oh wow but how is it referenced then 😮?

**Comment by BasThomas** [comment]
Ah yeah they've been working for quite a while I think 😮
